### PR TITLE
Transformation, add MergeJsonKeys API/functionality

### PR DESCRIPTION
**Pull request by EItanya**
# Description

Implement Transformation body parse MergeJsonKeys functionality recently introduced to `envoy-glow`

_Fill out any of the following sections that are relevant and remove the others_
## API changes
- Copied changes from envoy-gloo transformation proto for the MergeJsonKeys functionality

## Code changes
- Add conversion from gloo API to envoy API in transformation plugin
- Add unit test
- Add e2e test
# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: changelog/v1.18.0-beta11/bump-envoy.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+changelog:
+- type: DEPENDENCY_BUMP
+  dependencyOwner: solo-io
+  dependencyRepo: envoy-gloo
+  dependencyTag: v1.30.4-patch2
```
Comment (shashankram) [inline changelog/v1.18.0-beta11/bump-envoy.yaml:5]: The transformation plugin changes should be included in the changelog

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto [hunk 2]
```diff
@@ -293,6 +293,13 @@ message TransformationTemplate {
     // If you want to nest elements inside the body, use dot separator in the
     // extractor name.
     MergeExtractorsToBody merge_extractors_to_body = 6;
+    // A set of key-value pairs to merge into the JSON body.
+    // Each value will be rendered separately, and then placed into the JSON body at
+    // the specified key.
+    // There are a number of important caveats to using this feature:
+    // * This can only be used when the body is parsed as JSON.
+    // * This option does NOT work with advanced templates currently
```
Comment (shashankram) [inline projects/gloo/api/external/envoy/extensions/transformation/transformation.proto:301]: "advanced" is ambiguous. Maybe clarify what exactly won't work.

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto [hunk 3]
```diff
@@ -363,6 +370,33 @@ message Passthrough {}
...
+      "key1": "value1"
+    }
+    and the following MergeJsonKeys:
+    {
+      "key1": "{{ header("header1") }}",
+      "key2": "{{ header("header2") }}"
+    }
+    The resulting JSON body will be:
+    {
+      "key1": "header1_value",
```
Comment (shashankram) [inline projects/gloo/api/external/envoy/extensions/transformation/transformation.proto:393]: the existing key gets overwritten. That doesn't seem documented here

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto (hunk 2 above)
Comment (EItanya) [reply projects/gloo/api/external/envoy/extensions/transformation/transformation.proto:301]: advanced templates is a specific flag/concept to this API which has to be turned on, if a user sees this and is using advanced templates they will understand.

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto (hunk 3 above)
Comment (EItanya) [reply projects/gloo/api/external/envoy/extensions/transformation/transformation.proto:393]: Yes it is, `header1_value` is the output of the template above, the original value was `value1`

**Review by shashankram — DISMISSED** [review]

File: changelog/v1.18.0-beta11/bump-envoy.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+changelog:
+- type: DEPENDENCY_BUMP
```
Comment (nfuden) [inline changelog/v1.18.0-beta11/bump-envoy.yaml:2]: Should we also have a new_feature entry so customers know that this exists?

**Review by shashankram — DISMISSED** [review]

File: test/e2e/staged_transformation_test.go [hunk 5]
```diff
@@ -346,6 +347,57 @@ var _ = Describe("Staged Transformation", FlakeAttempts(3), func() {
 			}, "15s", ".5s").Should(Succeed())
 		})
 
+		It("Can modify specific body keys using MergeJsonKeys", func() {
```
Comment (nfuden) [inline test/e2e/staged_transformation_test.go:350]: Maybe 1 unhappy path.
Say specify nested path that only partially exists in the actual json

Comment (EItanya) [reply test/e2e/staged_transformation_test.go:350]: I got rid of nested path support, seemed like overkill and I didn't need it right now

File: changelog/v1.18.0-beta11/bump-envoy.yaml (hunk 4 above)
Comment (EItanya) [reply changelog/v1.18.0-beta11/bump-envoy.yaml:2]: 🤷 this feels like the kind of feature that really needs a doc anyway, so not sure that kind of discovery is valuable here.

**Review by jbohanon — DISMISSED** [review]

**Review by yuval-k — APPROVED** [review]

**Review by jbohanon — APPROVED** [review]

**Comment by nfuden** [comment]
🚢

**Pull request merged**